### ISSUE TITLE
RuntimeError: No suitable key found in keytab.

### ISSUE THREAD

**Issue by joakim-tjernlund** [openSUSE/cepces#21]
Moving https://github.com/ufven/cepces/issues/11 to here:
I get 3-4 of these each time I boot

2020-05-19 17:43:25,014 __main__:ERROR:Traceback (most recent call last):
  File "/usr/libexec/certmonger/cepces-submit", line 61, in main
    config = Configuration.load()
  File "/usr/lib64/python3.6/site-packages/cepces/config.py", line 131, in load
    return Configuration.from_parser(config)
  File "/usr/lib64/python3.6/site-packages/cepces/config.py", line 169, in from_parser
    return Configuration(endpoint, endpoint_type, cas, authn.handle())
  File "/usr/lib64/python3.6/site-packages/cepces/auth.py", line 94, in handle
    raise RuntimeError('No suitable key found in keytab.')
RuntimeError: No suitable key found in keytab.

Error is raised when there is no network yet.
Should it really do that or just wait a bit and then try again ?

**Comment by joakim-tjernlund** [comment]
> Honestly, this is probably a bug in the certmonger daemon startup of your distro. Certmonger should not be starting prior to the network being up. File a bug against your distro.

hmm, I don't think so. One should be able to inquire certmonger about cert like getcert list without network.
This feels more like missing error handling in cepces, it should not barf like that, just return an error to certmonger.

**Comment by dmulder** [comment]
Maybe, but the primary problem is your distro should not start certmonger before networking is up.

**Comment by dmulder** [comment]
And how would we fix this in cepces? Add a early check to make sure the network is reachable, then bail out? No, it's to OS job to ensure the network is up, not ours.

**Comment by dmulder** [comment]
We could catch the exception somewhere, and print a prettier message, and that might be more desirable for sysadmins, but it is not for those working on the code. I would much rather see the 'barf' so I can determine where things failed.

**Comment by dmulder** [comment]
> 
> One should be able to inquire certmonger about cert like getcert list without network.

OK, when I think about it, this part makes sense. We shouldn't break getcert just because the network is down.
@joakim-tjernlund could you test that out, just to be sure? With certificates already tracked, check if we break getcert if you disconnect the network.

**Comment by joakim-tjernlund** [comment]
> > One should be able to inquire certmonger about cert like getcert list without network.
> 
> OK, when I think about it, this part makes sense. We shouldn't break getcert just because the network is down. @joakim-tjernlund could you test that out, just to be sure? With certificates already tracked, check if we break getcert if you disconnect the network.

Looking deeper there seems to something else going on, I can see that NetworkManager starts long before certmonger and certmonger starts when NM becomes active later. Something else is bugging SOAPAuth:
```
 for principal in principals.strip().split('\n'):
            if realm:
                principal = '{}@{}'.format(principal, realm)

            try:
                auth = SOAPAuth.TransportKerberosAuthentication(
                    principal_name=principal,
                    init_ccache=ccache,
                    keytab=keytab,
                )
            except KerberosError:
                # Ignore
                pass

        if auth:
            return auth
        else:
            raise RuntimeError('No suitable key found in keytab.')
```

**Comment by c-erb** [comment]
Okay, I did some further testing with older Ubuntu versions and I managed to find two different things that can cause this issue.

1. Make sure the realm is configured correctly. You can configure it in the cepces configuration to be safe side. The difference between the Ubuntu versions is that the package krb5-config is finding the default realm in newer packaged versions and configures it in /etc/krb5.conf which is not the case for older versions.

2. Make sure your hostname is not longer than 15 characters. The issue here is that the configured principal names will not account for the computer object having a cut-off name if it exceeds 15 characters. So either you go with a shorter hostname or you change the principal names in the cepces configuration.

**Comment by joakim-tjernlund** [comment]
1. I have default_realm set in krb5.conf, should be enough?
2. Our hostname is FQDN, contains the domain as well. Other than that, hostname is less than 16 chars.

**Comment by c-erb** [comment]
> 1. I have default_realm set in krb5.conf, should be enough?
Yes that is enough.

> 2. Our hostname is FQDN, contains the domain as well. Other than that, hostname is less than 16 chars.
So is your output for `hostname` without domain and `hostname -f` with domain? Then this should be fine.
What you can do is check /etc/krb5.keytab (or wherever it might be on your system) to see which computerobject is addressed in there and if it fits the principalnames. 

Other than that I installed those packages on my test machine (additionally to what is included in the cepces setup:
  - libkrb5-dev
  - krb5-config
  - python3-pip
  - python3-requests-kerberos

**Comment by joakim-tjernlund** [comment]
> > 1. I have default_realm set in krb5.conf, should be enough?
> 
> Yes that is enough.
> 
> > 2. Our hostname is FQDN, contains the domain as well. Other than that, hostname is less than 16 chars.
> 
> So is your output for `hostname` without domain and `hostname -f` with domain? Then this should be fine. What you can do is check /etc/krb5.keytab (or wherever it might be on your system) to see which computerobject is addressed in there and if it fits the principalnames.

No, the other way around. hostname and hostname -f both have domain in it. hostname -s does not.
This I cannot change, something else breaks but dot remember what  ATM(maybe NFS?)

> 
> Other than that I installed those packages on my test machine (additionally to what is included in the cepces setup:
> 
>     * libkrb5-dev
> 
>     * krb5-config
> 
>     * python3-pip
> 
>     * python3-requests-kerberos

This is on Gentoo but I have there installed as well except for requests-kerberos, there I have requests-gssapi as later cepces has swithed but The problem was seen before that switch too.

**Comment by c-erb** [comment]
> No, the other way around. hostname and hostname -f both have domain in it. hostname -s does not. This I cannot change, something else breaks but dot remember what ATM(maybe NFS?)

If you checked the keytab file and the hostname is wrong there, than you could try to change the principals directly in the cepces config.

Like so, exchange hostname and domainname ofc.
host/HOSTNAME$$@DOMAINNAME
host/HOSTNAME$$
host/hostname$$.domainname$$@DOMAINNAME
HOSTNAME$$

**Comment by joakim-tjernlund** [comment]
> > No, the other way around. hostname and hostname -f both have domain in it. hostname -s does not. This I cannot change, something else breaks but dot remember what ATM(maybe NFS?)
> 
> If you checked the keytab file and the hostname is wrong there, than you could try to change the principals directly in the cepces config.
> 
> Like so, exchange hostname and domainname ofc. host/HOSTNAME$$@domainname host/HOSTNAME$$ host/hostname$$.domainname$$@domainname HOSTNAME$$

Not sure I follow, in keytab I have most variations:
```
HOST$@REALM
host/host.domain.com@REALM
host/HOST@REALM
```
and cepces principals are default:
```
principals=
  ${shortname}$$
  ${SHORTNAME}$$
  host/${SHORTNAME}
  host/${fqdn}
```
Seems to be match there

**Comment by joakim-tjernlund** [comment]
The error msg in cepces could be improved, right now it is just hardcoded:
   RuntimeError('No suitable key found in keytab.')
Including the error code from  SOAPAuth.TransportKerberosAuthentication() would help.

**Comment by joakim-tjernlund** [comment]
> Can someone check if [#47](https://github.com/openSUSE/cepces/pull/47) fixes it?

Patch does not apply on released SW

**Comment by joakim-tjernlund** [comment]
Handedited change into 0.3.9 an no, it still there:
```
journalctl -b | grep certmonger
Jun 25 10:14:19 se-jocke2-lx.infinera.com (rtmonger)[1608]: certmonger.service: Referenced but unset environment variable evaluates to an empty string: OPTS
Jun 25 10:14:19 se-jocke2-lx.infinera.com certmonger[1608]: 2025-06-25 10:14:19 [1608] Wrote to /var/lib/certmonger/requests/20250625080956
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]: 2025-06-25 10:14:20,231 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1731]: RuntimeError: No suitable key found in keytab.
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]: 2025-06-25 10:14:20,232 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1726]: RuntimeError: No suitable key found in keytab.
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]: 2025-06-25 10:14:20,232 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1725]: RuntimeError: No suitable key found in keytab.
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]: 2025-06-25 10:14:20,232 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1733]: RuntimeError: No suitable key found in keytab.
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]: 2025-06-25 10:14:20,232 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1723]: RuntimeError: No suitable key found in keytab.
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]: 2025-06-25 10:14:20,232 __main__:ERROR:Traceback (most recent call last):
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:   File "/usr/libexec/certmonger/cepces-submit", line 64, in main
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:     config = Configuration.load(global_overrides=global_overrides,
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 156, in load
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:     return Configuration.from_parser(config)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:   File "/usr/lib/python3.12/site-packages/cepces/config.py", line 196, in from_parser
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:     return Configuration(endpoint, endpoint_type, cas, authn.handle(), poll_interval, openssl_seclevel)
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:                                                        ^^^^^^^^^^^^^^
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:   File "/usr/lib/python3.12/site-packages/cepces/auth.py", line 100, in handle
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]:     raise RuntimeError('No suitable key found in keytab.')
Jun 25 10:14:20 se-jocke2-lx.infinera.com certmonger[1729]: RuntimeError: No suitable key found in keytab.
```

Maybe you can device some better logging ?
